### PR TITLE
Allow using arbitrary delimiters

### PR DESCRIPTION
**Pull request by Kijewski**
* Allow delimiters with arbitrary length
* Allow setting expr_start, etc. in Rust source

### CODE REVIEW

**Comment by Kijewski** [comment]
Actually, using a single character as delimiter (even such a character as newline) (w|sh)ould work, but then it would be difficult to extend the language in the future if we have to assume that any and all characters are used in the wild.

**Comment by djc** [comment]
What are you trying to achieve here? Conceptually I don't really like the idea of giving `Config` access to the `TemplateArgs`.

**Comment by Kijewski** [comment]
Well, 1st and foremost I wanted to remove the restriction that you have to use exactly two ASCII characters as delimiters. With this PR you can use any and all unicode characters.  And then I find setting up a config file cumbersome. Right now you can already set the whitespace handling in the code where you derive the template, without needing to resort to using a config file. So I added the option to set the other parts of the configuration, too. I pass &TemplateArgs instead of adding six new parameters, because until Rust adds named parameters having so many parameters is quite error prone, esp. when every parameter has the same type.

**Comment by Kijewski** [comment]
In the end I want to use askama to format a single latex file. Using braces as delimiters does not work very well in this context. Adding a config file would make the code less readable, IMO, because I have no code duplication anyway.

File: askama_derive/src/config.rs [hunk 1]
```diff
@@ -168,24 +168,29 @@ impl<'a> TryFrom<RawSyntax<'a>> for Syntax<'a> {
...
-        let ee = syntax.expr_start.as_bytes()[1];
-        if !((bs == cs && bs == es) || (be == ce && be == ee)) {
-            return Err(format!("bad delimiters block_start: {}, comment_start: {}, expr_start: {}, needs one of the two characters in common", syntax.block_start, syntax.comment_start, syntax.expr_start).into());
+        for (s1, s2) in [
+            (syntax.block_start, syntax.expr_start),
+            (syntax.block_start, syntax.comment_start),
+            (syntax.expr_start, syntax.comment_start),
+        ] {
+            if s1.starts_with(s2) || s2.starts_with(s1) {
+                return Err("A delimiter may not be the prefix of another delimiter".into());
```
Comment (djc) [inline askama_derive/src/config.rs:211]: Nit: I think our error message usually don't start with capitals.

**Review by djc — COMMENTED** [review]
Sorry for the slow feedback.

I think the goal here is good, but I still think that allowing `Config` to inspect the `TemplateArgs` is a layering violation. It should really be set up the other way around.

File: askama_derive/src/generator.rs [hunk 2]
```diff
@@ -163,37 +180,13 @@ impl TemplateArgs {
...
-                    return Err("whitespace value must be string literal".into());
-                }
-            } else {
+            } else if !capture_str_lit!(
+               "escape": args.escaping,
+               "ext": args.ext,
+               "syntax": args.syntax,
+               "config": args.config_path,
+               "whitespace": args.whitespace,
+            ) {
```
[the remark this answers is not in the document]
Comment (Kijewski) [reply askama_derive/src/generator.rs]: Yes, good idea!

File: askama_derive/src/config.rs (hunk 1 above)
Comment (Kijewski) [reply askama_derive/src/config.rs:211]: You're right. Will change.

**Comment by Kijewski** [comment]
> Sorry for the slow feedback.

No problem at all!

What do you think about adding a new syntax config struct `{whitespace, block_start, ..}` that I would pass around? Instead of `template_whitespace` in the current version, or the whole `TemplateArgs` I in my PR proposed?

**Comment by djc** [comment]
> What do you think about adding a new syntax config struct `{whitespace, block_start, ..}` that I would pass around? Instead of `template_whitespace` in the current version, or the whole `TemplateArgs` I in my PR proposed?

Sounds like a decent path forward? I'd have to see it in more detail to be sure, though.

**Comment by Kijewski** [comment]
PR needs to be redone entirely, because the codebase changed entirely.

**Pull request closed without merging**